gh-139038: Add JIT What's New for 3.15 (#142845)
python/cpython: Doc/whatsnew/3.15.rst

* :pep:`782`: :ref:`A new PyBytesWriter C API to create a Python bytes object
  <whatsnew315-pep782>`
* :ref:`Improved error messages <whatsnew315-improved-error-messages>`

> * :ref:`The JIT compiler has been significantly upgraded <whatsnew315-jit>`

Reviewer: Let's move it up one:
```suggestion
* :ref:`The JIT compiler has been significantly upgraded <whatsnew315-jit>`
* :ref:`Improved error messages <whatsnew315-improved-error-messages>`
```

* :pep:`782`: :ref:`A new PyBytesWriter C API to create a Python bytes object
  <whatsnew315-pep782>`
* :ref:`Improved error messages <whatsnew315-improved-error-messages>`
* The :ref:`JIT compiler <whatsnew315-jit>` has been significantly upgraded,

>   with 3-4% geometric mean performance improvement on the pyperformance benchmarks.

Reviewer 2: ```suggestion
  with 3-4% geometric mean performance improvement on the `pyperformance
  <https://github.com/python/pyperformance>`__ benchmarks.
```

* :pep:`782`: :ref:`A new PyBytesWriter C API to create a Python bytes object
  <whatsnew315-pep782>`
* :ref:`Improved error messages <whatsnew315-improved-error-messages>`
* The :ref:`JIT compiler <whatsnew315-jit>` has been significantly upgraded,
  with 3-4% geometric mean performance improvement on the `pyperformance

>   <https://github.com/python/pyperformance>`__ benchmarks.

Reviewer: This is a headline and should be brief. The details are linked below:
```suggestion
* The :ref:`JIT compiler <whatsnew315-jit>` has been significantly upgraded,
  with 3-4% geometric mean performance improvement
```

  (Contributed by Maurycy Pawłowski-Wieroński in :gh:`137628`.)

.. _whatsnew315-jit:

Upgraded JIT compiler
=====================

Results from the `pyperformance

>   <https://github.com/python/pyperformance>`__ benchmark suite report

Reviewer: ```suggestion
<https://github.com/python/pyperformance>`__ benchmark suite report
```

  (Contributed by Maurycy Pawłowski-Wieroński in :gh:`137628`.)

.. _whatsnew315-jit:

Upgraded JIT Compiler
=====================

The JIT compiler introduced in

> `Python 3.13 <https://docs.python.org/3.15/whatsnew/3.13.html#an-experimental-just-in-time-jit-compiler>`_

Reviewer 2: Can we use a `:ref:` here

> has been significantly upgraded.

Reviewer 2: It might be worth adding a little more colour to this lead sentence, currently it just restates the summary bullet. Words below are a sketch, please use something better :)
```suggestion
has been significantly upgraded, with a new internal implementation
and significant performance improvements.
```

Reviewer 3: I removed the sentence altogether.

> Current results from the pyperformance benchmark suite report

Reviewer 2: 'Current' always ages badly, perhaps just 'results'? Also, I don't think the para mentions what the base case is?

`3-4% <https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20251214-3.15.0a2%2B-6cddf04-JIT/bm-20251214-vultr-x86_64-python-6cddf04344a1e8ca9df5-3.15.0a2%2B-6cddf04-vs-base.svg>`_
geometric mean performance over the standard CPython interpreter built with
all optimizations enabled. The speedups range from roughly 20% slowdown to
over 100% speedup (ignoring the ``unpack_sequence`` microbenchmark) on the

> x86-64 Linux and AArch64 macOS systems. These results are not yet final.

Reviewer 2: Is this -20% - +100% with JIT vs no JIT on the same CPython commit, or versus a previous CPython with the JIT enabled?

Reviewer 3: JIT vs no JIT on the same CPython commit.

Reviewer 3: the preceding sentence says:
> over the standard CPython interpreter built with all optimizations enabled

> x86-64 Linux and AArch64 macOS systems. These results are not yet final.

Reviewer 2: Maybe put the not yet final note into an ``.. attention::`` admonition?

The major upgrades to the JIT are as follows:

* LLVM 21 build-time dependency
* New tracing frontend
* Basic register allocation in the JIT
* More JIT optimizations
* Better machine code generation

> * ...watch this space for more to come before 3.15 beta!

Reviewer 2: ```suggestion
```

Reviewer 3: aww :(

LLVM 21 Build-Time Dependency

> -----------------------------

Reviewer 2: Can we combine these paragraphs into the bullet list above?

Reviewer 3: I would prefer not to. That makes it harder to click on the left hand bar and navigate to the relevant section. Also it removes their prominence when navigating.

Reviewer 2: The challenge is that *What's New* is already incredibly long, and I'd prefer not to have lots of these sub-headings throughout -- it will make the sidebar even harder to use, especially if others copy the subheading per paragraph approach. Appreciate that we want the improvements to have prominence, but it's a balancing act on usability of the document as a whole.

A

Reviewer 3: Right. Is there some way to have a subsection that does not show up in the side bar?

Reviewer 2: Perhaps [`.. rubric::`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-rubric)?

Reviewer 3: that's exactly it, thank you!

`3-4% <https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20251214-3.15.0a2%2B-6cddf04-JIT/bm-20251214-vultr-x86_64-python-6cddf04344a1e8ca9df5-3.15.0a2%2B-6cddf04-vs-base.svg>`_
geometric mean performance improvement for the JIT over the standard CPython
interpreter built with all optimizations enabled. The speedups for JIT
builds versus no JIT builds range from roughly 20% slowdown to over
100% speedup (ignoring the ``unpack_sequence`` microbenchmark) on
x86-64 Linux and AArch64 macOS systems.

.. attention::
    These results are not yet final.

The major upgrades to the JIT are as follows:

* LLVM 21 build-time dependency
* New tracing frontend
* Basic register allocation in the JIT
* More JIT optimizations
* Better machine code generation

.. rubric:: LLVM 21 Build-Time Dependency

The JIT compiler now uses LLVM 21 for build-time stencil generation. As
always, LLVM is only needed when building CPython with the JIT enabled;
end users running Python do not need LLVM installed. Instructions for
installing LLVM can be found in the
`JIT compiler documentation <Tools/jit/README.md>`_ for all supported

> platforms.

Reviewer 2: This link won't work on d.p.o. (see the rendered preview)
```suggestion
installing LLVM can be found in the JIT compiler documentation
(:file:`Tools/jit/README.md`) for all supported platforms.
```

(Contributed by Savannah Ostrowski in :gh:`140973`.)

.. rubric:: A New Tracing Frontend

The JIT compiler now supports significantly more bytecode operations and
control flow than in Python 3.14, enabling speedups on a wider variety of
code. For example, simple Python object creation is now understood by the
3.15 JIT compiler. Overloaded operations and generators are also partially
supported. This was made possible by an overhauled JIT tracing frontend
that records actual execution paths through code, rather than estimating
them as the previous implementation did.

(Contributed by Ken Jin in :gh:`139109`. Support for Windows added by
Mark Shannon :gh:`141703`.)

.. rubric:: Basic Register Allocation in the JIT

A basic form of register allocation has been added to the JIT compiler's
optimizer. This allows the JIT compiler to avoid certain stack operations
altogether and instead operate on registers. This allows the JIT to produce
more efficient traces by avoiding reads and writes to memory.

(Contributed by Mark Shannon in :gh:`135379`.)

.. rubric:: More JIT Optimizations

> More `constant-propagation <https://en.wikipedia.org/wiki/Constant_folding>`_

Reviewer 2: We recently changed [our guidance](https://devguide.python.org/documentation/markup/#hyperlinks), we should use double underscore (anonymous links) in new text
```suggestion
More `constant-propagation <https://en.wikipedia.org/wiki/Constant_folding>`__
```

is now performed. This means when the JIT compiler detects that certain user
code results in constants, the code can be simplified by the JIT.

> (Contributed by Ken Jin and Savannah Ostrowski in :gh:`132732`.)

Reviewer 2: consistent whitespace (+ two more)
```suggestion

(Contributed by Ken Jin and Savannah Ostrowski in :gh:`132732`.)
```

The JIT avoids :term:`reference count`\ s where possible. This generally
reduces the cost of most operations in Python.

> (Contributed by Ken Jin, Donghee Na, Nadeshiko Manju, Savannah Ostrowski,

Reviewer 2: ```suggestion

(Contributed by Ken Jin, Donghee Na, Nadeshiko Manju, Savannah Ostrowski,
```

Noam Cohen, Tomas Roun, PuQing in :gh:`134584`.)

.. rubric:: Better Machine Code Generation

The JIT compiler's machine code generator now produces better machine code
for x86-64 and AArch64 macOS and Linux targets. In general, users should
experience lower memory usage for generated machine code and more efficient
machine code versus the old JIT.

> (Contributed by Brandt Bucher in :gh:`136528` and :gh:`136528`.

Reviewer 2: ```suggestion

(Contributed by Brandt Bucher in :gh:`136528` and :gh:`136528`.
```

  (Contributed by Maurycy Pawłowski-Wieroński in :gh:`137628`.)

.. _whatsnew315-jit:

Upgraded JIT Compiler

> =====================

Reviewer: Use sentence case for headers:

https://devguide.python.org/documentation/style-guide/#capitalization
```suggestion
Upgraded JIT compiler
=====================
```

> Results from the pyperformance benchmark suite report

Reviewer: Link here instead of in the headline:
```suggestion
Results from the `pyperformance
  <https://github.com/python/pyperformance>`__ benchmark suite report
```

> `3-4% <https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20251214-3.15.0a2%2B-6cddf04-JIT/bm-20251214-vultr-x86_64-python-6cddf04344a1e8ca9df5-3.15.0a2%2B-6cddf04-vs-base.svg>`_

Reviewer: ```suggestion
`3-4% <https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20251214-3.15.0a2%2B-6cddf04-JIT/bm-20251214-vultr-x86_64-python-6cddf04344a1e8ca9df5-3.15.0a2%2B-6cddf04-vs-base.svg>`__
```

geometric mean performance improvement for the JIT over the standard CPython
interpreter built with all optimizations enabled. The speedups for JIT
builds versus no JIT builds range from roughly 20% slowdown to over
100% speedup (ignoring the ``unpack_sequence`` microbenchmark) on
x86-64 Linux and AArch64 macOS systems.

.. attention::
    These results are not yet final.

> The major upgrades to the JIT are as follows:

Reviewer: ```suggestion
The major upgrades to the JIT are:
```

* LLVM 21 build-time dependency
* New tracing frontend
* Basic register allocation in the JIT
* More JIT optimizations
* Better machine code generation

> .. rubric:: LLVM 21 Build-Time Dependency

Reviewer: ```suggestion
.. rubric:: LLVM 21 build-time dependency
```

The JIT compiler now uses LLVM 21 for build-time stencil generation. As
always, LLVM is only needed when building CPython with the JIT enabled;
end users running Python do not need LLVM installed. Instructions for
installing LLVM can be found in the JIT compiler documentation

> (:file:`Tools/jit/README.md`) for all supported platforms.

Reviewer: ```suggestion
installing LLVM can be found in the `JIT compiler documentation
<https://github.com/python/cpython/blob/main/Tools/jit/README.md>`__
for all supported platforms.
```

(Contributed by Savannah Ostrowski in :gh:`140973`.)

> .. rubric:: A New Tracing Frontend

Reviewer: ```suggestion
.. rubric:: A new tracing frontend
```

The JIT compiler now supports significantly more bytecode operations and
control flow than in Python 3.14, enabling speedups on a wider variety of
code. For example, simple Python object creation is now understood by the
3.15 JIT compiler. Overloaded operations and generators are also partially
supported. This was made possible by an overhauled JIT tracing frontend
that records actual execution paths through code, rather than estimating
them as the previous implementation did.

(Contributed by Ken Jin in :gh:`139109`. Support for Windows added by

> Mark Shannon :gh:`141703`.)

Reviewer: ```suggestion
Mark Shannon in :gh:`141703`.)
```

> .. rubric:: Basic Register Allocation in the JIT

Reviewer: ```suggestion
.. rubric:: Basic register allocation in the JIT
```

A basic form of register allocation has been added to the JIT compiler's
optimizer. This allows the JIT compiler to avoid certain stack operations
altogether and instead operate on registers. This allows the JIT to produce
more efficient traces by avoiding reads and writes to memory.

(Contributed by Mark Shannon in :gh:`135379`.)

> .. rubric:: More JIT Optimizations

Reviewer: ```suggestion
.. rubric:: More JIT optimizations
```

More `constant-propagation <https://en.wikipedia.org/wiki/Constant_folding>`__
is now performed. This means when the JIT compiler detects that certain user
code results in constants, the code can be simplified by the JIT.

(Contributed by Ken Jin and Savannah Ostrowski in :gh:`132732`.)

The JIT avoids :term:`reference count`\ s where possible. This generally
reduces the cost of most operations in Python.

(Contributed by Ken Jin, Donghee Na, Nadeshiko Manju, Savannah Ostrowski,
Noam Cohen, Tomas Roun, PuQing in :gh:`134584`.)

> .. rubric:: Better Machine Code Generation

Reviewer: ```suggestion
.. rubric:: Better machine code generation
```
